### PR TITLE
refactor: arguments handling of replicas

### PR DESCRIPTION
**Pull request by JoanFM**

### CODE REVIEW

File: jina/enums.py [hunk 1]
```diff
@@ -238,6 +238,7 @@ class PeaRoleType(BetterEnum):
     HEAD = 1
     TAIL = 2
     PARALLEL = 3
+    # this is confusing, a PEA is not a REPLICA, it would make more sense to have this as Pod type
```
Comment (florian-hoenicke) [inline jina/enums.py:241]: agree, this should be removed.

File: jina/peapods/pods/compoundpod.py [hunk 2]
```diff
@@ -112,9 +119,8 @@ def all_args(
         :return: arguments for all Peas and pods
         """
         args = {
-            'peas': ([self.replicas_args['head']] if self.replicas_args['head'] else [])
-            + ([self.replicas_args['tail']] if self.replicas_args['tail'] else []),
-            'replicas': self.replicas_args['replicas'],
+            'peas': [self.head_args, self.tail_args],
```
Comment (florian-hoenicke) [inline jina/peapods/pods/compoundpod.py:122]: you are right, the head and the tail always exist. so we can simplify this

File: jina/peapods/pods/compoundpod.py [hunk 3]
```diff
@@ -29,11 +29,12 @@ def __init__(
     ):
         super().__init__(args, needs)
         self.replica_list = []  # type: List['Pod']
-        if isinstance(args, Dict):
-            # This is used when a Pod is created in a remote context, where peas & their connections are already given.
-            self.replicas_args = args
-        else:
-            self.replicas_args = self._parse_args(args)
+        # we will see how to have `CompoundPods` in remote later when we add tests for it
```
Comment (florian-hoenicke) [inline jina/peapods/pods/compoundpod.py:32]: I like that we don't consider args being a dictionary

File: jina/peapods/pods/compoundpod.py [hunk 4]
```diff
@@ -124,9 +125,9 @@ def all_args(
         :return: arguments for all Peas and pods
         """
         args = {
-            'peas': ([self.replicas_args['head']] if self.replicas_args['head'] else [])
-            + ([self.replicas_args['tail']] if self.replicas_args['tail'] else []),
-            'replicas': self.replicas_args['replicas'],
+            'head': self.head_args,
```
Comment (florian-hoenicke) [inline jina/peapods/pods/compoundpod.py:128]: this is cleaner - I like

**Review by florian-hoenicke — DISMISSED** [review]

File: jina/peapods/pods/compoundpod.py [hunk 5]
```diff
@@ -29,11 +29,12 @@ def __init__(
     ):
         super().__init__(args, needs)
         self.replica_list = []  # type: List['Pod']
-        if isinstance(args, Dict):
-            # This is used when a Pod is created in a remote context, where peas & their connections are already given.
-            self.replicas_args = args
-        else:
-            self.replicas_args = self._parse_args(args)
+        # we will see how to have `CompoundPods` in remote later when we add tests for it
+        self.head_args = self._parse_head_args(args)
```
Comment (JoanFM) [inline jina/peapods/pods/compoundpod.py:33]: ok I need to set this `parse_head_args`

File: jina/peapods/pods/compoundpod.py [hunk 6]
```diff
@@ -137,11 +74,14 @@ def num_peas(self) -> int:
 
         :return: total number of peas including head and tail
         """
-        return sum([replica.num_peas for replica in self.replica_list]) + len(self.peas)
+        return sum([replica.num_peas for replica in self.replicas]) + 2
```
Comment (bwanglzu) [inline jina/peapods/pods/compoundpod.py:77]: +2 means head and tail pea?

Comment (JoanFM) [reply jina/peapods/pods/compoundpod.py:77]: yes

Comment (JoanFM) [reply jina/peapods/pods/compoundpod.py:77]: `CompoundPod` is a special kind of Pod containing many Pod inside and surrounded by `head` and `tail`.

This one will ALWAYS have Head and Tail

**Review by florian-hoenicke — APPROVED** [review]

**Pull request merged**